MREUnityRuntime/MREUnityRuntimeLib/Core/Components/AnimationComponent.cs

```diff
@@ -122,128 +196,267 @@ void AddKeyframe(MWAnimationKeyframe keyframe)
                     clip.SetCurve("", kv.Value.Type, kv.Key, kv.Value.Curve);
                 }
 
-                clip.AddEvent(new AnimationEvent()
+                _animationData[animationName] = new AnimationData()
                 {
-                    functionName = "AnimationEndedEvent",
-                    time = clip.length
-                });
+                    IsInternal = isInternal,
+                    OnCompleteCallback = onCompleteCallback
+                };
+
+                float initialTime = 0f;
+                float initialSpeed = 1f;
+                bool initialEnabled = false;
+
+                if (initialState != null)
+                {
+                    initialTime = initialState.Time ?? initialTime;
+                    initialSpeed = initialState.Speed ?? initialSpeed;
+                    initialEnabled = initialState.Enabled ?? initialEnabled;
+                }
 
                 var animation = GetOrCreateUnityAnimationComponent();
                 animation.AddClip(clip, animationName);
 
-                callback?.Invoke();
+                SetAnimationState(animationName, initialTime, initialSpeed, initialEnabled);
+
+                onCreatedCallback?.Invoke();
             });
 
             continuation.Start();
         }
 
-        internal void StartAnimation(string animationName, float? animationTime, bool? paused, bool? hasRootMotion)
+        internal void Interpolate(
+            ActorPatch finalFrame,
+            string animationName,
+            float duration,
+            float[] curve,
+            bool enabled,
+            Action onCompleteCallback)
         {
-            var animation = GetOrLookUpUnityAnimationComponent();
-            if (animation != null)
+            // Ensure duration is in range [0...n].
+            duration = Math.Max(0, duration);
+
+            // If the curve is malformed, fallback to linear.
+            if (curve.Length != 4)
             {
-                if (animation[animationName] != null)
+                curve = new float[] { 0, 0, 1, 1 };
+            }
+
+            // Are we patching the transform?
+            bool animateTransform = finalFrame.Transform != null && finalFrame.Transform.IsPatched();
+            var finalTransform = finalFrame.Transform;
+
+            // What parts of the transform are we animating?
+            bool animatePosition = animateTransform && finalTransform.Position != null && finalTransform.Position.IsPatched();
+            bool animateRotation = animateTransform && finalTransform.Rotation != null && finalTransform.Rotation.IsPatched();
+            bool animateScale = animateTransform && finalTransform.Scale != null && finalTransform.Scale.IsPatched();
+
+            // Ensure we have a well-formed rotation quaternion.
+            if (animateRotation)
+            {
+                bool wellFormed =
```

Comment:
Quaternions must be normalized, or Unity will start spitting errors. Can we try to parse as a Quaternion as a validation test?

Reply:
Good idea, will do.

Reply:
Bummer, our version of of Unity doesn't have a Quaternion.Normalize method. Is introduced in 2018.3.

Reply:
Updated to ensure quaternion is normalized.